internal/tokenstore/token_store.go

```diff
@@ -0,0 +1,236 @@
+package tokenstore
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/config"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/gwerrors"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/models"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/oidc"
+	"github.com/SwissDataScienceCenter/renku-gateway/internal/sessions"
+	"github.com/redis/go-redis/v9"
+)
+
+type TokenStore struct {
+	ExpiryMarginMinutes int
+
+	providerStore oidc.ClientStore
+	tokenRepo     models.TokenRepository
+}
+
+func (ts TokenStore) ExpiryMargin() time.Duration {
+	return time.Duration(ts.ExpiryMarginMinutes) * time.Minute
+}
+
+func (ts *TokenStore) GetFreshAccessToken(ctx context.Context, tokenID string) (models.AuthToken, error) {
+	token, err := ts.tokenRepo.GetAccessToken(ctx, tokenID)
+	if err != nil {
+		if err == redis.Nil {
+			return models.AuthToken{}, gwerrors.ErrTokenNotFound
+		} else {
+			return models.AuthToken{}, err
+		}
+	}
+
+	if token.ExpiresSoon(ts.ExpiryMargin()) {
+		slog.Debug(
+			"TOKEN STORE",
+			"message",
+			"access token expires soon",
+			"token",
+			token.String(),
+		)
+		newTokenSet, err := ts.refreshAccessToken(ctx, token)
+		if err != nil {
+			slog.Info(
+				"TOKEN STORE",
+				"message",
+				"refreshAccessToken failed, will try to reload the token",
+				"token",
+				token.String(),
+			)
+			reloadedToken, err := ts.tokenRepo.GetAccessToken(ctx, tokenID)
+			if err != nil {
+				token = reloadedToken
+			}
```

Comment:
Fixed in b2eed526e1bdc4968bcf6a93eca3d08f1c32d8fb (set `token = reloadedToken` when err is nil).

The token is reloaded from redis, so that we have a refreshed token if it was refreshed by another instance.

Reply:
> Why not just fail here?

If the token is still valid, we can continue using it, even if refreshing fails.